Allow null types

veronicagg opened the pull request:
Addressing https://github.com/Azure/oav/issues/155
@amarzavery please take  a look when you get a chance, I'll be adding tests, but wanted to send a PR with the modified code, since we talked about this. 
I changed the condition on when nullable applies a bit, so want to know if you agree or we should be doing it differently. 
Basically, I noticed that we have some specs that specifically set x-nullable to false, in which case I think that even if the property is not required, we should probably honor that setting. 
Logic is now that if x-nullable is not undefined then we honor the value, and if the property is undefined, then we look at whether the property is required or not. Thoughts?

amarzavery commented:
@veronicagg - Please resolve conflicts with upstream master

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,6 +580,39 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
   return model;
 }
 
+exports.allowNullType = function allowNullType(entity, isPropRequired) {
```
The valueType could be a primitive type or an inline object with properties or a reference to another object (composite type) or an array type.
- If it is a reference to another type then it is fine.
- If it is an inline object with properties then we need to process it // allowNullableTypes()
- If it is a primitive type like string, number, boolean then we need to process it // allowNullTypes()
- If it is an array then we need to process it

veronicagg commented:
@amarzavery I've updated the code to account for the cases we've talked about and added tests, let me know what you think.
I see the CI is failing for other tests, I need to look into that, I wonder if my merge didn't work well, as I see something failing for x-ms-parameterized-host.

amarzavery commented:
@veronicagg - Please set
`options.shouldResolveNullableTypes = false` for semantic validation over [here](https://github.com/Azure/oav/blob/master/lib/validate.js#L68)

amarzavery commented on lib/commands/resolve-spec.js:
```diff
@@ -53,6 +53,12 @@ exports.builder = {
   },
+  n: {
+    alias: 'nullable',
+    describe: 'Should nullable be resolved?',
```
Should nullable **types** be resolved?

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
+    }
+
+    let isPropNullable = entity['x-nullable'] && typeof entity['x-nullable'] === 'boolean';
+    if ((isPropNullable === undefined && !isPropRequired) || isPropNullable) {
```
It is worthwhile providing a comment or may be putting a True False logic table that we had drawn on the white board in the comments over here. That will be helpful to us in the future.

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
+
+    let isPropNullable = entity['x-nullable'] && typeof entity['x-nullable'] === 'boolean';
+    if ((isPropNullable === undefined && !isPropRequired) || isPropNullable) {
+      let savedEntity = entity
```
// Handling nullable parameters

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
+        delete entity.type
+
+      }
+      else {
```
as per JS code style, else should be on the same line as the if closing bracket
```
} else {
 . ..
}
```

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
+      if (savedEntity.in) {
+        entity.oneOf = [{ "type": entity.type }, { "type": "null" }];
+        delete entity.type
+
```
remove extra line

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
+
+  // if there's a $ref
+  if (entity && entity["$ref"]) {
     let isPropNullable = entity['x-nullable'] && typeof entity['x-nullable'] === 'boolean';
```
duplicate code, can we extract this into a function and call that function?

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
+      if (modelProperties[propName].properties || modelProperties[propName].additionalProperties) {
         modelProperties[propName] = exports.allowNullableTypes(modelProperties[propName]);
       }
       else {
```
else on the same line

amarzavery commented on lib/util/utils.js:
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
+    if (model.type == 'array') {
+      if (model.items) {
+        // if items object contains inline properties
+        if (model.items.properties) {
```
what about model.items.additionalProperties?

amarzavery commented on lib/validators/specResolver.js:
```diff
@@ -621,13 +621,26 @@ class SpecResolver {
     for (let defName in definitions) {
       let model = definitions[defName];
-      utils.allowNullableTypes(model);
+      definitions[defName] = utils.allowNullableTypes(model);
```
wouldn't this make the definition itself a `oneOf` of that type and null?

veronicagg commented:
Still pending some code review feedback, and full validation on azure-rest-api-specs repo.

amarzavery commented:
everything LGTM. Bump the version in package.json and also update the changelog.md

amarzavery commented:
This [StackOverflow link](https://stackoverflow.com/questions/47374980/schema-object-without-a-type-attribute-in-swagger-2-0) provides good explanation about requiredness `type` in a json document.
